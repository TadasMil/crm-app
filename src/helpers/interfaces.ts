export interface IUser {
  firstName: string
  lastName: string
  email: string
  city: string
  street: string
  houseNumber: string
  zipCode: string
}

export interface IUserEntity extends IUser {
  coordinates: Object
  id: string
}
