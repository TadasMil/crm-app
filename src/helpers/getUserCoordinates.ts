export const getUserCoordinates = async (address: string) => {
  const query = new URLSearchParams({
    location: address || '',
    key: import.meta.env.VITE_GEOCODING_API_KEY!.toString(),
  }).toString()

  const path = 'http://www.mapquestapi.com/geocoding/v1/address?' + query

  return await fetch(path, {
    method: 'GET',
  }).then((res) =>
    res.json().then((data) => {
      const [location] = data.results
      const [addressLocation] = location.locations

      return { lat: addressLocation.latLng }
    })
  )
}
