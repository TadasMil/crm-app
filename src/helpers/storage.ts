import { IUserEntity } from './interfaces'

const LOCAL_STORAGE_KEYS = {
  USERS: 'crm-users',
}

export const getObject = (key: string) => {
  const value = localStorage.getItem(key)
  return value ? JSON.parse(value) : []
}

export const setObject = (key: string, value: any) => {
  if (value) {
    localStorage.setItem(key, JSON.stringify(value))
  } else {
    localStorage.removeItem(key)
  }
}

const getUsers = () => getObject(LOCAL_STORAGE_KEYS.USERS)

const addNewUser = (user: IUserEntity) => {
  const users = [...getUsers(), user]
  setObject(LOCAL_STORAGE_KEYS.USERS, users)
}

export const storage = {
  getUsers,
  addNewUser,
}
