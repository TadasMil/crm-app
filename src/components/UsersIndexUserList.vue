<script setup name="UsersIndexUserList" lang="ts">
import { IUserEntity } from '../helpers/interfaces'

const props = defineProps<{
  users: IUserEntity[]
}>()
</script>

<template>
  <div class="relative overflow-x-auto shadow-md sm:rounded-lg">
    <table class="w-full text-left text-sm">
      <thead class="bg-primary text-xs uppercase text-white">
        <tr>
          <th scope="col" class="px-6 py-3">Full Name</th>
          <th scope="col" class="px-6 py-3">Email</th>
          <th scope="col" class="px-6 py-3">City</th>
          <th scope="col" class="px-6 py-3">Street</th>
          <th scope="col" class="px-6 py-3">House Number</th>
          <th scope="col" class="px-6 py-3">Zipcode</th>
          <th scope="col" class="px-6 py-3">
            <span class="sr-only">Edit</span>
          </th>
        </tr>
      </thead>
      <tbody v-if="users.length">
        <tr
          v-for="user in users"
          :key="user.id"
          class="border-b odd:bg-white even:bg-gray-50 dark:border-gray-700 dark:bg-gray-800 odd:dark:bg-gray-800 even:dark:bg-gray-700"
        >
          <th
            scope="row"
            class="whitespace-nowrap px-6 py-4 font-medium text-gray-900 dark:text-white"
          >
            {{ user.firstName }} {{ user.lastName }}
          </th>
          <td class="px-6 py-4">{{ user.email }}</td>
          <td class="px-6 py-4">{{ user.city }}</td>
          <td class="px-6 py-4">{{ user.street }}</td>
          <td class="px-6 py-4">{{ user.houseNumber }}</td>
          <td class="px-6 py-4">{{ user.zipCode }}</td>
          <td class="px-6 py-4 text-right">
            <router-link
              :to="{ name: 'UserEdit', params: { id: user.id } }"
              class="font-medium text-primary hover:underline"
            >
              Edit
            </router-link>
          </td>
        </tr>
      </tbody>
      <tbody v-else class="h-20 w-full bg-white">
        <tr class="text-md text-center">
          <td colspan="7">No users found.</td>
        </tr>
      </tbody>
    </table>
  </div>
</template>
