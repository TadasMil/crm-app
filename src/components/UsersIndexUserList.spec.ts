import { render } from '@testing-library/vue'
import { IUserEntity } from 'src/helpers/interfaces'
import { describe, expect, it } from 'vitest'

import UsersIndexUserList from './UsersIndexUserList.vue'

describe('UsersIndexUserList.spec.js', () => {
  it('should render the component', () => {
    const user: IUserEntity = {
      firstName: 'Tadas123',
      lastName: 'Tadas12',
      email: 'ga@test.com',
      city: 'Vilnius',
      street: 'Zirmunu',
      houseNumber: '141',
      zipCode: 'LT-06612',
      id: '2adff068-f113-45d0-9a95-145d7b2d24e8',
      coordinates: { lat: { lat: 54.723772, lng: 25.297623 } },
    }

    const { html } = render(UsersIndexUserList, {
      props: {
        users: [user],
      },
      global: {
        stubs: ['router-link'],
      },
    })

    expect(html()).toMatchSnapshot()
  })

  it('should render the no data component given empty users list', () => {
    const { html } = render(UsersIndexUserList, {
      props: {
        users: [],
      },
      global: {
        stubs: ['router-link'],
      },
    })

    expect(html()).toMatchSnapshot()
  })
})
