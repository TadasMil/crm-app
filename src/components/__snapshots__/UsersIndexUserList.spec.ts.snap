// Vitest Snapshot v1

exports[`UsersIndexUserList.spec.js > should render the component 1`] = `
"<div class=\\"relative overflow-x-auto shadow-md sm:rounded-lg\\">
  <table class=\\"w-full text-left text-sm\\">
    <thead class=\\"bg-primary text-xs uppercase text-white\\">
      <tr>
        <th scope=\\"col\\" class=\\"px-6 py-3\\">Full Name</th>
        <th scope=\\"col\\" class=\\"px-6 py-3\\">Email</th>
        <th scope=\\"col\\" class=\\"px-6 py-3\\">City</th>
        <th scope=\\"col\\" class=\\"px-6 py-3\\">Street</th>
        <th scope=\\"col\\" class=\\"px-6 py-3\\">House Number</th>
        <th scope=\\"col\\" class=\\"px-6 py-3\\">Zipcode</th>
        <th scope=\\"col\\" class=\\"px-6 py-3\\"><span class=\\"sr-only\\">Edit</span></th>
      </tr>
    </thead>
    <tbody>
      <tr class=\\"border-b odd:bg-white even:bg-gray-50 dark:border-gray-700 dark:bg-gray-800 odd:dark:bg-gray-800 even:dark:bg-gray-700\\">
        <th scope=\\"row\\" class=\\"whitespace-nowrap px-6 py-4 font-medium text-gray-900 dark:text-white\\">Tadas123 Tadas12</th>
        <td class=\\"px-6 py-4\\">ga@test.com</td>
        <td class=\\"px-6 py-4\\">Vilnius</td>
        <td class=\\"px-6 py-4\\">Zirmunu</td>
        <td class=\\"px-6 py-4\\">141</td>
        <td class=\\"px-6 py-4\\">LT-06612</td>
        <td class=\\"px-6 py-4 text-right\\">
          <router-link-stub to=\\"[object Object]\\" class=\\"font-medium text-primary hover:underline\\"></router-link-stub>
        </td>
      </tr>
    </tbody>
  </table>
</div>"
`;

exports[`UsersIndexUserList.spec.js > should render the no data component given empty users list 1`] = `
"<div class=\\"relative overflow-x-auto shadow-md sm:rounded-lg\\">
  <table class=\\"w-full text-left text-sm\\">
    <thead class=\\"bg-primary text-xs uppercase text-white\\">
      <tr>
        <th scope=\\"col\\" class=\\"px-6 py-3\\">Full Name</th>
        <th scope=\\"col\\" class=\\"px-6 py-3\\">Email</th>
        <th scope=\\"col\\" class=\\"px-6 py-3\\">City</th>
        <th scope=\\"col\\" class=\\"px-6 py-3\\">Street</th>
        <th scope=\\"col\\" class=\\"px-6 py-3\\">House Number</th>
        <th scope=\\"col\\" class=\\"px-6 py-3\\">Zipcode</th>
        <th scope=\\"col\\" class=\\"px-6 py-3\\"><span class=\\"sr-only\\">Edit</span></th>
      </tr>
    </thead>
    <tbody class=\\"h-20 w-full bg-white\\">
      <tr class=\\"text-md text-center\\">
        <td colspan=\\"7\\">No users found.</td>
      </tr>
    </tbody>
  </table>
</div>"
`;
